Sources/RDeviceIdentifier/RDeviceIdentifier.swift

```diff
@@ -0,0 +1,283 @@
+    ///          different provisioning profiles, will not produce the same device identifier.
+    ///
+    /// - Returns: A string uniquely identifying the device the application is currently running on.
+    ///         If the keychain is not available (i.e. the device is locked) and no value has been
+    ///         retrieved yet, `nil` is returned and the developer should try again when the device
+    ///         is unlocked.
+    ///
+    /// - Precondition: NSObjectInaccessibleException if the application's entitlements do not include the
+    ///            access group required to access the device identifier. See @ref device-information-keychain-setup "Setting up the keychain"
+    ///            for more information.
+    /// - Precondition: NSInternalInconsistencyException if the application is misconfigured and the first
+    ///            access group does not match the application's bundle identifier. See @ref device-information-keychain-setup "Setting up the keychain"
+    ///            for more information.
+    @objc public class var uniqueDeviceIdentifier: String? {
+        // If we already have a value, return it.
+        if let _uniqueDeviceIdentifier = _uniqueDeviceIdentifier {
+            return _uniqueDeviceIdentifier
+        }
+
+        // First, try to grab the application identifier prefix (=bundle seed it)
+        // and build the access group from it.
+        var accessGroup: String?
+        do {
+            if let result = try keychain.addIfNonExistent(),
+               let defaultAccessGroup = result[String(kSecAttrAccessGroup)] as? String,
+               let teamName = defaultAccessGroup.components(separatedBy: ".").first {
+                accessGroup = "\(teamName).\(keychainAccessGroup)"
+
+                // While we're at it, why not check developers didn't do the unthinkable?
+                checkKeychainAccessGroupOrder(defaultAccessGroup)
+            }
+        } catch {
+            // Keychain is not available
+            return nil
+        }
+
+        // Try to clean things up
+        keychain.clear()
+
+        // Try to find the device identifier in the keychain.
+        // Here we always have a bundle seed id.
+        do {
+            if let accessGroup = accessGroup, let uniqueDeviceIdData = try keychain.search(for: accessGroup) {
+                // Device id found!
+                _uniqueDeviceIdentifier = uniqueDeviceIdData.hexString
+                return _uniqueDeviceIdentifier
+            }
+        } catch {
+            // Keychain problem
+            return nil
+        }
+
+        // Get identifierForVendor and write it to the keychain.
+        // If it succeeds, then assign the result to '_deviceIdData'.
+        var deviceIdData: Data?
+        var idForVendor = UIDevice.current.identifierForVendor?.uuidString ?? UUID().uuidString
+        // Filter out nil, empty, or zeroed strings (e.g. "00000000-0000-0000-0000-000000000000")
```

Comment:
Is it worth to add some logging in that case? or maybe even assertion? I'm not sure. What do you guys think?

Reply:
Added a handle for that case